README.md
![Inky pHAT](inky-phat-logo.png)
http://shop.pimoroni.com/products/inky-phat

**Note** that Inky pHAT uses a new display now (as of early December 2017), and **requires an updated library**. This new library automagically detects the display version, so should work interchangeably with the old and new displays.

## Installing

### Full install (recommended):

We've created an easy installation script that will install all pre-requisites and get your Inky pHAT
up and running with minimal efforts. To run it, fire up Terminal which you'll find in Menu -> Accessories -> Terminal
on your Raspberry Pi desktop, as illustrated below:

![Finding the terminal](http://get.pimoroni.com/resources/github-repo-terminal.png)

In the new terminal window type the command exactly as it appears below (check for typos) and follow the on-screen instructions:

```bash
curl https://get.pimoroni.com/inkyphat | bash
```

If you choose to download examples you'll find them in `/home/pi/Pimoroni/inkyphat/`.

### Manual install:

#### Library install for Python 3:

```bash
sudo apt-get install python3-inkyphat
```

#### Library install for Python 2:

```bash
sudo apt-get install python-inkyphat
```

### Development:

If you want to contribute, or like living on the edge of your seat by having the latest code, you should clone this repository, `cd` to the library directory, and run:

```bash
sudo apt-get install python-dev python-setuptools
sudo python3 setup.py install
```
(or `sudo python setup.py install` whichever your primary Python environment may be)

In all cases you will have to enable the SPI bus.

## Documentation & Support

* Guides and tutorials - https://learn.pimoroni.com/inky-phat
* Function reference - http://docs.pimoroni.com/inkyphat/
* GPIO Pinout - http://pinout.xyz/pinout/inky_phat
* Get help - http://forums.pimoroni.com/c/support


packaging/CHANGELOG
inkyphat (1.0.1) stable; urgency=low

  * BugFix: Fixed colour support issues with V1 Inky pHAT

 -- Phil Howard <phil@pimoroni.com>  Tue, 16 Oct 2018 00:00:00 +0000

inkyphat (1.0.0) stable; urgency=low

  * Feature: Added support for Black/Yellow and Black

 -- Phil Howard <phil@pimoroni.com>  Fri, 06 Jul 2018 00:00:00 +0000

inkyphat (0.1.1) stable; urgency=low

  * Removed GPIO.cleanup call for Inky pHAT v2, so subsequent display updates will work

 -- Phil Howard <phil@pimoroni.com>  Thu, 07 Dec 2017 00:00:00 +0000

inkyphat (0.1.0) stable; urgency=low

  * Added support for new display on Inky pHAT v2

 -- Phil Howard <phil@pimoroni.com>  Thu, 07 Dec 2017 00:00:00 +0000

inkyphat (0.0.2) stable; urgency=low

  * Tweaked LCD init to run every update
  * Added LCD deinit to switch off driver IC
  * Added clear method

 -- Phil Howard <phil@pimoroni.com>  Thu, 13 Jul 2017 00:00:00 +0000

inkyphat (0.0.1) stable; urgency=low

  * Initial release

 -- Phil Howard <phil@pimoroni.com>  Fri, 23 Jun 2017 00:00:00 +0000


examples/weather.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import glob
import json
import time
import urllib
import sys
from PIL import Image, ImageFont

try:
    import requests
except ImportError:
    exit("This script requires the requests module\nInstall with: sudo pip install requests")

import inkyphat


print("""Inky pHAT: Weather

Displays weather information for a given location. The default location is Sheffield-on-Sea.

""")

if len(sys.argv) < 2:
    print("""Usage: {} <colour>
       Valid colours: red, yellow, black
""".format(sys.argv[0]))
    sys.exit(0)

colour = sys.argv[1]

try:
    inkyphat.set_colour(colour)
except ValueError:
    print('Invalid colour "{}" for V{}\n'.format(colour, inkyphat.get_version()))
    if inkyphat.get_version() == 2:
        sys.exit(1)
    print('Defaulting to "red"')

inkyphat.set_border(inkyphat.BLACK)

CITY = "Sheffield"
COUNTRYCODE = "GB"
WARNING_TEMP = 25.0


def get_location():
    res = requests.get('http://ipinfo.io')
    if(res.status_code == 200):
        json_data = json.loads(res.text)
        return json_data
    return {}

# Python 2 vs 3 breaking changes.
def encode(qs):
    val = ""
    try:
        val = urllib.urlencode(qs).replace("+","%20")
    except:
        val = urllib.parse.urlencode(qs).replace("+", "%20")
    return val

def get_weather(address):
    base = "https://query.yahooapis.com/v1/public/yql?"
    query = "select * from weather.forecast where woeid in (select woeid from geo.places(1) where text=\""+address+"\")"
    qs={"q": query, "format": "json", "env": "store://datatables.org/alltableswithkeys"}

    uri = base + encode(qs)                                        

    res = requests.get(uri)
    if(res.status_code==200):
        json_data = json.loads(res.text)
        return json_data

    return {}

icons = {}
masks = {}

location_string = "{city}, {countrycode}".format(city=CITY, countrycode=COUNTRYCODE)

weather = get_weather(location_string)

icon_map = {
	"snow": [5, 6, 7, 8, 10, 13, 14, 15, 16, 17, 18, 41, 42, 43, 46],
	"rain": [9, 11, 12],
	"cloud": [19, 20, 21, 22, 25, 26, 27, 28, 29, 30, 44],
	"sun": [32, 33, 34, 36],
	"storm": [0, 1, 2, 3, 4, 37, 38, 39, 45, 47],
        "wind": [23, 24]
}

pressure = 0
temperature = 0
weather_icon = None

if "channel" in weather["query"]["results"]:
    results = weather["query"]["results"]["channel"]
    pressure = results["atmosphere"]["pressure"]
    temperature = (int(results["item"]["forecast"][0]["high"]) - 32) * .5556
    code = int(results["item"]["forecast"][0]["code"])

    for icon in icon_map:
        if code in icon_map[icon]:
            weather_icon = icon
            break

else:
    print("Warning, no weather information found!")

# Load our icon files and generate masks
for icon in glob.glob("resources/icon-*.png"):
    icon_name = icon.split("icon-")[1].replace(".png", "")
    icon_image = Image.open(icon)
    icons[icon_name] = icon_image
    masks[icon_name] = inkyphat.create_mask(icon_image)

# Load the built-in FredokaOne font
font = ImageFont.truetype(inkyphat.fonts.FredokaOne, 22)

# Load our backdrop image
inkyphat.set_image("resources/backdrop.png")

# Let's draw some lines!
inkyphat.line((69, 36, 69, 81)) # Vertical line
inkyphat.line((31, 35, 184, 35)) # Horizontal top line
inkyphat.line((69, 58, 174, 58)) # Horizontal middle line
inkyphat.line((169, 58, 169, 58), 2) # Red seaweed pixel :D

# And now some text

datetime = time.strftime("%d/%m %H:%M")

inkyphat.text((36, 12), datetime, inkyphat.WHITE, font=font)

inkyphat.text((72, 34), "T", inkyphat.WHITE, font=font)
inkyphat.text((92, 34), u"{:.2f}°".format(temperature), inkyphat.WHITE if temperature < WARNING_TEMP else inkyphat.RED, font=font)

inkyphat.text((72, 58), "P", inkyphat.WHITE, font=font)
inkyphat.text((92, 58), "{}".format(pressure), inkyphat.WHITE, font=font)

# Draw the current weather icon over the backdrop
if weather_icon is not None:
    inkyphat.paste(icons[weather_icon], (28, 36), masks[weather_icon])

else:
    inkyphat.text((28, 36), "?", inkyphat.RED, font=font)

# And show it!
inkyphat.show()
